### PR TITLE
4.2 - Docs: Workflow Beta

### PR DESCRIPTION
**Pull request by benarent**
This PR has been opened against a new `4.2-base` branch to make reviewing easier.  It outlines the new features for Workflows.

### CODE REVIEW

**Review by fspmarshall — COMMENTED** [review]
Points that came up during the final review of #3131 and should probably be documented:

- The `options.max_session_ttl` field of the *requested* role (`dba` in your examples) provides an upper-limit on how long the access-request will be valid.  If, for example, we wanted to ensure that `contractor` would need to re-request `dba` every hour, setting the `options.max_session_ttl` field of `dba` to 1 hour would be the correct way to do this.

- Granting a role with administrative abilities could allow a user to *permanently* upgrade their privileges (e.g. if `contractor` was granted `admin` for some reason).  This isn't particularly surprising, but it should be mentioned explicitly.  The `deny.request` block can help mitigate the risk of doing this by accident.

File: docs/4.2/cli-docs.md [hunk 1]
```diff
@@ -481,6 +481,60 @@ tctl users rm sally,tim
...
+Approve a users request. **Usage**: `tctl request approve [token]`
+
+### Arguments
+
+* `<tokens>` - comma-separated list of Teleport tokens.
+
+### Examples
+
+``` bsh
+tctl request approve bc8ca931-fec9-4b15-9a6f-20c13c5641a9,3iwoxHenTp4zNttjWvn5niL5oGNvCnBormFanU
```
[the remark this answers is not in the document]
Comment (benarent) [reply docs/4.2/cli-docs.md]: Thanks for the clarification, I've gone ahead and just duplicated it.

File: docs/4.2/enterprise/index.md [hunk 2]
```diff
@@ -102,3 +102,104 @@ In addition, Teleport checks if the binary was compiled against an approved
...
+!!! warning "Warning: Workflows are currently in Alpha"
+    This feature is currently in alpha. 
+    If you have a question please post to our [community](https://community.gravitational.com/), or file bugs on [Github](https://github.com/gravitational/teleport/issues/new). 
+
+With Teleport 4.2 we've introduced the ability for users to request additional roles. The workflow API makes it easy to dynamically approve or deny these requests.
+
+## Setup
+
+**Contractor Role**
+This role let's the contractor request the role DBA.
```
Comment (webvictim) [inline docs/4.2/enterprise/index.md]: let's -> lets

File: docs/4.2/enterprise/index.md [hunk 3]
```diff
@@ -102,3 +102,104 @@ In addition, Teleport checks if the binary was compiled against an approved
...
+  allow:
+    request:
+      roles: ['dba']
+    # ...
+  deny:
+    # ...
+```
+
+**DBA Role**
+This role let's the contractor request the role DBA.
```
Comment (webvictim) [inline docs/4.2/enterprise/index.md]: let's -> lets

File: docs/4.2/enterprise/index.md [hunk 4]
```diff
@@ -102,3 +102,104 @@ In addition, Teleport checks if the binary was compiled against an approved
...
+**DBA Role**
+This role let's the contractor request the role DBA. 
+```yaml
+kind: role
+metadata:
+  name: dba
+spec:
+  options:
+    # ...
+    # Only provides the contractor to use this role for 1hr from time of request.
```
Comment (webvictim) [inline docs/4.2/enterprise/index.md]: `Only provides` -> `Only allows`

File: docs/4.2/enterprise/index.md [hunk 5]
```diff
@@ -102,3 +102,104 @@ In addition, Teleport checks if the binary was compiled against an approved
...
+    options.max_session_ttl: 1hr
+  allow:
+    # ...
+  deny:
+    # ...
+```
+
+**Admin Role**
+This role let's the admin approve the contractors request.
```
Comment (webvictim) [inline docs/4.2/enterprise/index.md]: let's -> lets

File: docs/4.2/enterprise/index.md [hunk 6]
```diff
@@ -102,3 +102,104 @@ In addition, Teleport checks if the binary was compiled against an approved
...
+      verbs: [list, read, update, delete]
+```
+
+
+```bash
+$ tsh login teleport-cluster --request-roles=dba
+Seeking request approval... (id: bc8ca931-fec9-4b15-9a6f-20c13c5641a9)
+```
+
+As a Teleport Administrator.
```
Comment (webvictim) [inline docs/4.2/enterprise/index.md]: `Administrator:`

File: docs/4.2/cli-docs.md [hunk 7]
```diff
@@ -481,6 +481,60 @@ tctl users rm sally,tim
...
+
+### Examples
+
+``` bsh
+tctl request deny request-id-1, request-id-2
+```
+
+## tctl request rm 
+
+Delete a users role request. **Usage**: `tctl request rm [token]`
```
Comment (webvictim) [inline docs/4.2/cli-docs.md]: `users` -> `user's`

File: docs/4.2/cli-docs.md [hunk 8]
```diff
@@ -481,6 +481,60 @@ tctl users rm sally,tim
...
+
+### Examples
+
+``` bsh
+tctl request approve request-id-1, request-id-2
+```
+
+## tctl request deny 
+
+Denies a users request. **Usage**: `tctl request deny [token]`
```
Comment (webvictim) [inline docs/4.2/cli-docs.md]: `users` -> `user's`

File: docs/4.2/cli-docs.md [hunk 9]
```diff
@@ -481,6 +481,60 @@ tctl users rm sally,tim
...
+
+``` bsh
+tctl request ls
+Token                                Requestor Metadata       Created At (UTC)    Status  
+------------------------------------ --------- -------------- ------------------- ------- 
+request-id-1                         alice     roles=dictator 07 Nov 19 19:38 UTC PENDING
+```
+## tctl request approve 
+
+Approve a users request. **Usage**: `tctl request approve [token]`
```
Comment (webvictim) [inline docs/4.2/cli-docs.md]: `users` -> `user's`

**Review by webvictim — COMMENTED** [review]
Sorry, I saw this late - some spelling/grammar nits.

I can re-add them on the final PR review if needed.

**Comment by benarent** [comment]
Thanks Gus, I've gone ahead and fixed these.

**Pull request merged**